Add regression test information to COMPILING.md

TGWDB opened the pull request:
This adds information on how to run regression tests using bnoth cmake and make to COMPILING.md.
 Also ensures COMPILING.md follows clang-format-7.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

hannes-steffenhagen-diffblue commented on REGRESSION_TESTS.md:
```diff
@@ -0,0 +1,43 @@
+# CMAKE
+
+To build with release flags in cmake use
+`-DCMAKE_BUILD_TYPE=Release` as an argument to the cmake command.
```
This is unrelated to running the regression tests (it's already mentioned in `COMPILING.md` at any rate).

TGWDB replied:
Agree it is redundant. Removed.

tautschnig replied:
That's already documented in COMPILING.md?!

hannes-steffenhagen-diffblue commented on COMPILING.md:
```diff
@@ -255,6 +255,10 @@ require manual modification of build files.
 
+   Note that if you do not specify the `build` directory the current directory
+   will be the build target. This will modify many version controlled files
+   and be messy to fix.
```
I think I'll just add a change to cmake to make this impossible.

TGWDB replied:
Handled by #5782

tautschnig commented on COMPILING.md:
```diff
@@ -255,6 +255,10 @@ require manual modification of build files.
    to generate IDE projects by supplying the `-G` flag.  Run `cmake -G` for a
    comprehensive list of supported back-ends.
 
+   Note that if you do not specify the `build` directory the current directory
+   will be the build target. This will modify many version controlled files
+   and be messy to fix.
```
Rather than adding documentation I'd prefer one of the options discussed in https://stackoverflow.com/questions/1208681/with-cmake-how-would-you-disable-in-source-builds/10306476#10306476 to disallow this behaviour.

hannes-steffenhagen-diffblue replied:
@tautschnig I added https://github.com/diffblue/cbmc/pull/5782 for this.

TGWDB replied:
Removed as this is now superseded by PR #5782.

tautschnig commented on REGRESSION_TESTS.md:
```diff
@@ -0,0 +1,41 @@
+This document assumes you have already been able to build CPROVER on
+your chosen architecture. For details on building CPROVER see
+COMPILING.md.
```
Could we please merge this documentation with doc/architectural/compilation-and-development.md rather than adding another file?

TGWDB replied:
Happy to merge this file with another one. I started a draft pull request to make sure I wasn't putting bad information in or adding things in the wrong place. If we all agree on doc/architectural/compilation-and-development.md as the right location I'll move everything there.

tautschnig replied:
Thank you for bringing up all these gaps in our documentation, and coming up with fixes for these! I'd just like to make sure we don't have the documentation in multiple places. I'm no longer able to tell what's actually needed for someone to get started, and where they might be looking for documentation. Is perhaps `CONTRIBUTING.md` (which we don't currently have) the canonical place for people to get started?

TGWDB replied:
I'm happy to try and sort it all out. I'm using myself as a test subject for this (before I know my way around well), so hopefully do not create too many (draft) proposals to change things that are explained somewhere.

For (a possible) `CONTRIBUTING.md`, would this also include the various links and information on policies for style and committing?

tautschnig replied:
If we had a `CONTRIBUTING.md`, perhaps it could serve as a collection of pointers towards documentation, pointing people to `COMPILING.md`, `CODING_STANDARD.md`, and the sections of the manual (such as `doc/architectural/compilation-and-development.md`).

peterschrammel replied:
Btw, the documentation is published at http://cprover.diffblue.com (see link on top fo README)

TGWDB replied:
What is our link preference here, should we link to http://cprover.diffblue.com/ or do we reference inside the repository since that page is generated from doc/architectural/front-page.md ?

tautschnig replied:
`README.md` references http://cprover.diffblue.com/, so I'd go with that in top-level files (`README.md`, `COMPILING.md`, `CODING_STANDARD.md`, and possibly any files to be added at that level).

hannes-steffenhagen-diffblue commented:
@tautschnig This CSmith check just failed: https://github.com/diffblue/cbmc/pull/5781/checks?check_run_id=1762881802

How do we interpret that?

tautschnig commented:
> @tautschnig This CSmith check just failed: https://github.com/diffblue/cbmc/pull/5781/checks?check_run_id=1762881802
> 
> How do we interpret that?

See #5780 - hoping for @kroening or @chrisr-diffblue to review this ASAP.

TGWDB committed:
Add regression test information to COMPILING.md

This adds information on how to run regression tests
using bnoth cmake and make to COMPILING.md.
Also ensures COMPILING.md follows clang-format-7.

chrisr-diffblue commented on COMPILING.md:
```diff
@@ -482,3 +482,42 @@ successfully on Windows or macOS.
    compiled binary will be placed in `cbmc/src/cbmc/cbmc`.
+This document assumes you have already been able to build CPROVER on
+your chosen architecture. For details on building CPROVER see
+COMPILING.md.
```
Since this content is _in_ `COMPILING.md` - maybe its better to say `see section BLAH` ?

TGWDB replied:
Fixed

chrisr-diffblue commented on COMPILING.md:
````diff
@@ -482,3 +482,42 @@ successfully on Windows or macOS.
+simply running make as follows.
+```
+cd regression
+make
````
I believe there's a slight difference between cmake/make in what tests actually run, which might be worth either clarifying, or adjusting the instructions to elide. That is around unit tests... I believe the `ctest` commands run _both_ regression _and_ unit tests, whereas I though the make command you show here _only_ runs the regression tests. My recommendation might be to tweak the section title to say "RUNNING REGRESSION AND UNIT TESTS", and then tweak the instructions for make to show how to run unit tests as well?

TGWDB replied:
I think this is clearer now.

chrisr-diffblue reviewed (approved):
Thanks, thats a great addition to the docs!
